File: scripts/qa/test_adblock_module_filter.py
#!/usr/bin/env python3
"""Regression: PROMAX line-split vs full vs skip."""
import os
import sys

SCRIPT_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.insert(0, SCRIPT_DIR)

from pipeline.adblock_manager import (  # noqa: E402
    AMPLIFY_NEXUS_DIR,
    AdBlockManager,
    LOCAL_MODULES_DIR,
)
from hub.promax_line_classifier import (
    classify_promax_line,
    module_ingest_mode,
    should_keep_promax_line,
)
from hub.promax_module_split import split_module_sections


def test_filename_gate() -> None:
    mgr = AdBlockManager()
    cases = [
        (os.path.join(LOCAL_MODULES_DIR, "知乎去广告.sgmodule"), "split"),
        (os.path.join(LOCAL_MODULES_DIR, "扫描全能王解锁.sgmodule"), "skip"),
        (os.path.join(AMPLIFY_NEXUS_DIR, "BiliBili.Enhanced.sgmodule"), "skip"),
        (os.path.join(AMPLIFY_NEXUS_DIR, "扫描全能王解锁.sgmodule"), "skip"),
        (os.path.join(AMPLIFY_NEXUS_DIR, "WeChat_Enhance.sgmodule"), "skip"),
    ]
    for path, expected in cases:
        if not os.path.isfile(path):
            continue
        text = open(path, encoding="utf-8").read()
        assert module_ingest_mode(path, text) == expected, path


def test_bilibili_bundle_split_keeps_ad_only() -> None:
    path = os.path.join(
        AMPLIFY_NEXUS_DIR, "📺 BiliBili增强合集.sgmodule"
    )
    if not os.path.isfile(path):
        return
    text = open(path, encoding="utf-8").read()
    ad_sections, stats = split_module_sections(text, source_path=path)
    assert stats["kept_lines"] > 0
    assert stats["kept_lines"] < stats["total_lines"]
    scripts = ad_sections.get("Script", [])
    assert any("ADBlock" in s or "helper" in s.lower() for s in scripts)
    assert not any("BiliBili.Enhanced" in s for s in scripts)
    assert not any("BiliBili.Global" in s for s in scripts)
    rules = ad_sections.get("Rule", [])
    assert any("REJECT" in r for r in rules)


def test_pdd_module_full_ingest_keeps_all_body_rewrite() -> None:
    path = os.path.join(LOCAL_MODULES_DIR, "拼多多去广告.sgmodule")
    if not os.path.isfile(path):
        return
    text = open(path, encoding="utf-8").read()
    assert module_ingest_mode(path, text) == "full"
    jq_count = text.count("http-response-jq")
    _, stats = split_module_sections(text, source_path=path)
    assert stats["skipped_lines"] == 0
    ad_sections, _ = split_module_sections(text, source_path=path)
    assert len(ad_sections.get("Body Rewrite", [])) == jq_count


def test_weibo_ad_scripts_kept_in_split() -> None:
    path = os.path.join(LOCAL_MODULES_DIR, "🐦 微博去广告合集.sgmodule")
    if not os.path.isfile(path):
        return
    line = (
        "微博热搜页面广告 = type=http-response, pattern=^https?:\\/\\/m?api\\.weibo\\.c(n|om)\\/2\\/"
        "(page|flowpage)\\?, script-path=https://raw.githubusercontent.com/fmz200/wool_scripts/main/"
        "Scripts/weibo/weibo_ads.js, requires-body=true"
    )
    assert classify_promax_line(line, "Script", source_path=path) == "ad"


def test_zhihu_legacy_body_rewrite_kept() -> None:
    line = (
        r'http-response ^https:\/\/api\.zhihu\.com\/search\/recommend_query\/v2\? '
        r'"recommend_queries":\{.+\} "recommend_queries":{}'
    )
    assert classify_promax_line(line, "Body Rewrite") == "ad"


def test_rednote_scripts_full_module() -> None:
    path = os.path.join(LOCAL_MODULES_DIR, "RedNote.sgmodule")
    if not os.path.isfile(path):
        return
    text = open(path, encoding="utf-8").read()
    assert module_ingest_mode(path, text) == "full"
    _, stats = split_module_sections(text, source_path=path)
    assert stats["skipped_lines"] == 0


def test_unlock_weibo_script_still_dropped() -> None:
    path = os.path.join(LOCAL_MODULES_DIR, "🐦 微博去广告合集.sgmodule")
    line = (
        "解锁微博会员APP图标 = type=http-response, pattern=^https?:\\/\\/new\\.vip\\.weibo\\.c(n|om)\\/"
        "aj\\/appicon\\/list, script-path=https://example/unlock.js"
    )
    assert not should_keep_promax_line(line, "Script", source_path=path)


def test_pdd_bottom_tabs_body_rewrite_kept() -> None:
    line = (
        "http-response-jq ^https:\\/\\/api\\.pinduoduo\\.com\\/api\\/alexa\\/homepage\\/hub\\? "
        "'.result.bottom_tabs? |= map(select(.link | IN(\"index.html\", \"chat_list.html\", \"personal.html\")))'"
    )
    assert classify_promax_line(line, "Body Rewrite") == "ad"


def test_classifier_adblock_script() -> None:
    line = (
        "📺 BiliBili.ADBlock.feed.index.request.json = type=http-request,"
        "script-path=https://github.com/BiliUniverse/ADBlock/releases/download/v0.6.24/request.bundle.js"
    )
    assert classify_promax_line(line, "Script") == "ad"
    enh = (
        "📺 BiliBili.Enhanced.x.resource.show.tab.v2 = type=http-response,"
        "script-path=https://github.com/BiliUniverse/Enhanced/releases/download/v0.5.13/response.bundle.js"
    )
    assert classify_promax_line(enh, "Script") == "enhance"


def test_functional_resolve_includes_split_bundle() -> None:
    mgr = AdBlockManager()
    paths, _ = mgr.load_functional_source_paths()
    by_path = {p: m for p, m in paths}
    bundle = os.path.join(AMPLIFY_NEXUS_DIR, "📺 BiliBili增强合集.sgmodule")
    if os.path.isfile(bundle):
        assert by_path.get(os.path.normpath(bundle)) == "split"
    assert not any("BiliBili.Enhanced.sgmodule" in p for p in by_path)
    assert not any("WeChat_Enhance" in p for p in by_path)


def test_localmodules_only_functional_scan() -> None:
    mgr = AdBlockManager()
    paths, _ = mgr.load_functional_source_paths()
    joined = "\n".join(p for p, _m in paths)
    assert "narrow_pierce" not in joined
    assert "LocalModules" in joined or any("LocalModules" in p for p, _ in paths)


def test_amplify_nexus_not_auto_scanned() -> None:
    mgr = AdBlockManager()
    paths, _ = mgr.load_functional_source_paths()
    joined = "\n".join(p for p, _m in paths)
    assert "iRingo" not in joined
    assert "YouTube.Enhance" not in joined
    assert "Apple服务增强" not in joined


def test_sukka_adblock_allowlisted_full() -> None:
    from pipeline.adblock_manager import LOCAL_SOURCES_DIR

    path = os.path.join(
        LOCAL_SOURCES_DIR, "[Sukka] Enhance Better ADBlock for Surge.sgmodule"
    )
    if not os.path.isfile(path):
        return
    mgr = AdBlockManager()
    assert mgr.resolve_module_ingest_mode(path) == "full"


if __name__ == "__main__":
    test_pdd_module_full_ingest_keeps_all_body_rewrite()
    test_weibo_ad_scripts_kept_in_split()
    test_zhihu_legacy_body_rewrite_kept()
    test_rednote_scripts_full_module()
    test_unlock_weibo_script_still_dropped()
    test_pdd_bottom_tabs_body_rewrite_kept()
    test_classifier_adblock_script()
    test_filename_gate()
    test_bilibili_bundle_split_keeps_ad_only()
    test_functional_resolve_includes_split_bundle()
    test_localmodules_only_functional_scan()
    test_amplify_nexus_not_auto_scanned()
    test_sukka_adblock_allowlisted_full()
    print("test_adblock_module_filter: OK")

File: create/qa/test_adblock_module_filter.go
package qa

import (
	"fmt"
	"path/filepath"
	"strings"

	"github.com/nyamiiko/script_hub/go_scripts/hub"
	"github.com/nyamiiko/script_hub/go_scripts/pipeline"
)

func runAdblockTest(name string, f func() error) error {
	if err := f(); err != nil {
		return fmt.Errorf("%s: %v", name, err)
	}
	return nil
}

func testFilenameGate() error {
	cases := []struct {
		path     string
		expected string
	}{
		{filepath.Join(hub.MODULE_LOCAL_DIR, "知乎去广告.sgmodule"), "split"},
		{filepath.Join(hub.MODULE_LOCAL_DIR, "扫描全能王解锁.sgmodule"), "skip"},
		{filepath.Join(hub.SURGE_AMPLIFY_NEXUS_DIR, "BiliBili.Enhanced.sgmodule"), "skip"},
		{filepath.Join(hub.SURGE_AMPLIFY_NEXUS_DIR, "扫描全能王解锁.sgmodule"), "skip"},
		{filepath.Join(hub.SURGE_AMPLIFY_NEXUS_DIR, "WeChat_Enhance.sgmodule"), "skip"},
	}

	for _, c := range cases {
		if !hub.ValidateFileExists(c.path, "") {
			continue
		}
		text := hub.ReadFileString(c.path)
		if mode := hub.ModuleIngestMode(c.path, text); mode != c.expected {
			return fmt.Errorf("expected %s, got %s for %s", c.expected, mode, c.path)
		}
	}
	return nil
}

func testBilibiliBundleSplitKeepsAdOnly() error {
	path := filepath.Join(hub.SURGE_AMPLIFY_NEXUS_DIR, "📺 BiliBili增强合集.sgmodule")
	if !hub.ValidateFileExists(path, "") {
		return nil
	}
	text := hub.ReadFileString(path)
	adSections, stats := hub.SplitModuleSections(text, path)

	if stats.KeptLines <= 0 {
		return fmt.Errorf("stats.KeptLines <= 0")
	}
	if stats.KeptLines >= stats.TotalLines {
		return fmt.Errorf("stats.KeptLines >= stats.TotalLines")
	}

	var scripts []string
	for name, lines := range adSections {
		if name == "Script" {
			scripts = append(scripts, lines...)
		}
	}

	hasAd := false
	for _, s := range scripts {
		if strings.Contains(s, "ADBlock") || strings.Contains(strings.ToLower(s), "helper") {
			hasAd = true
		}
		if strings.Contains(s, "BiliBili.Enhanced") || strings.Contains(s, "BiliBili.Global") {
			return fmt.Errorf("found BiliBili.Enhanced/Global in Script section")
		}
	}
	if !hasAd {
		return fmt.Errorf("no ADBlock or helper found in Script section")
	}

	hasReject := false
	for name, lines := range adSections {
		if name == "Rule" {
			for _, r := range lines {
				if strings.Contains(r, "REJECT") {
					hasReject = true
				}
			}
		}
	}
	if !hasReject {
		return fmt.Errorf("no REJECT found in Rule section")
	}
	return nil
}

func testPddModuleFullIngestKeepsAllBodyRewrite() error {
	path := filepath.Join(hub.MODULE_LOCAL_DIR, "拼多多去广告.sgmodule")
	if !hub.ValidateFileExists(path, "") {
		return nil
	}
	text := hub.ReadFileString(path)
	if hub.ModuleIngestMode(path, text) != "full" {
		return fmt.Errorf("expected full mode")
	}
	jqCount := strings.Count(text, "http-response-jq")
	adSections, stats := hub.SplitModuleSections(text, path)
	if stats.SkippedLines != 0 {
		return fmt.Errorf("stats.SkippedLines != 0")
	}
	bodyRewriteCount := 0
	for name, lines := range adSections {
		if name == "Body Rewrite" {
			bodyRewriteCount += len(lines)
		}
	}
	if bodyRewriteCount != jqCount {
		return fmt.Errorf("bodyRewriteCount != jqCount")
	}
	return nil
}

func testWeiboAdScriptsKeptInSplit() error {
	path := filepath.Join(hub.MODULE_LOCAL_DIR, "🐦 微博去广告合集.sgmodule")
	if !hub.ValidateFileExists(path, "") {
		return nil
	}
	line := "微博热搜页面广告 = type=http-response, pattern=^https?:\\/\\/m?api\\.weibo\\.c(n|om)\\/2\\/(page|flowpage)\\?, script-path=https://raw.githubusercontent.com/fmz200/wool_scripts/main/Scripts/weibo/weibo_ads.js, requires-body=true"
	if hub.ClassifyPromaxLine(line, "Script", path) != "ad" {
		return fmt.Errorf("expected ad")
	}
	return nil
}

func testZhihuLegacyBodyRewriteKept() error {
	line := `http-response ^https:\/\/api\.zhihu\.com\/search\/recommend_query\/v2\? "recommend_queries":\{.+\} "recommend_queries":{}`
	if hub.ClassifyPromaxLine(line, "Body Rewrite", "") != "ad" {
		return fmt.Errorf("expected ad")
	}
	return nil
}

func testRednoteScriptsFullModule() error {
	path := filepath.Join(hub.MODULE_LOCAL_DIR, "RedNote.sgmodule")
	if !hub.ValidateFileExists(path, "") {
		return nil
	}
	text := hub.ReadFileString(path)
	if hub.ModuleIngestMode(path, text) != "full" {
		return fmt.Errorf("expected full mode")
	}
	_, stats := hub.SplitModuleSections(text, path)
	if stats.SkippedLines != 0 {
		return fmt.Errorf("stats.SkippedLines != 0")
	}
	return nil
}

func testUnlockWeiboScriptStillDropped() error {
	path := filepath.Join(hub.MODULE_LOCAL_DIR, "🐦 微博去广告合集.sgmodule")
	line := "解锁微博会员APP图标 = type=http-response, pattern=^https?:\\/\\/new\\.vip\\.weibo\\.c(n|om)\\/aj\\/appicon\\/list, script-path=https://example/unlock.js"
	if hub.ShouldKeepPromaxLine(line, "Script", path) {
		return fmt.Errorf("expected ShouldKeepPromaxLine to return false")
	}
	return nil
}

func testPddBottomTabsBodyRewriteKept() error {
	line := `http-response-jq ^https:\/\/api\.pinduoduo\.com\/api\/alexa\/homepage\/hub\? '.result.bottom_tabs? |= map(select(.link | IN("index.html", "chat_list.html", "personal.html")))'`
	if hub.ClassifyPromaxLine(line, "Body Rewrite", "") != "ad" {
		return fmt.Errorf("expected ad")
	}
	return nil
}

func testClassifierAdblockScript() error {
	line := "📺 BiliBili.ADBlock.feed.index.request.json = type=http-request,script-path=https://github.com/BiliUniverse/ADBlock/releases/download/v0.6.24/request.bundle.js"
	if hub.ClassifyPromaxLine(line, "Script", "") != "ad" {
		return fmt.Errorf("expected ad for %s", line)
	}
	enh := "📺 BiliBili.Enhanced.x.resource.show.tab.v2 = type=http-response,script-path=https://github.com/BiliUniverse/Enhanced/releases/download/v0.5.13/response.bundle.js"
	if hub.ClassifyPromaxLine(enh, "Script", "") != "enhance" {
		return fmt.Errorf("expected enhance for %s", enh)
	}
	return nil
}

func testFunctionalResolveIncludesBilibiliAdblockLocal() error {
	mgr := pipeline.NewAdBlockManager()
	paths, _ := mgr.LoadFunctionalSourcePaths()
	byPath := make(map[string]string)
	for _, p := range paths {
		byPath[p[0]] = p[1]
	}

	localAdblock := filepath.Clean(filepath.Join(hub.ROOT, "rulesets/Sources", "BiliBili.ADBlock.sgmodule"))
	if hub.ValidateFileExists(localAdblock, "") {
		if byPath[localAdblock] != "full" {
			return fmt.Errorf("expected full mode for local adblock")
		}
	}
	for p := range byPath {
		if strings.Contains(p, "BiliBili.Enhanced.sgmodule") {
			return fmt.Errorf("found BiliBili.Enhanced.sgmodule")
		}
		if strings.Contains(p, "WeChat_Enhance") {
			return fmt.Errorf("found WeChat_Enhance")
		}
	}
	return nil
}

func testLocalModulesOnlyFunctionalScan() error {
	mgr := pipeline.NewAdBlockManager()
	paths, _ := mgr.LoadFunctionalSourcePaths()
	for _, p := range paths {
		if strings.Contains(p[0], "narrow_pierce") {
			return fmt.Errorf("found narrow_pierce")
		}
	}
	hasLocalModules := false
	for _, p := range paths {
		if strings.Contains(p[0], "LocalModules") {
			hasLocalModules = true
		}
	}
	if !hasLocalModules && len(paths) > 0 {
		return fmt.Errorf("LocalModules not scanned")
	}
	return nil
}

func testAmplifyNexusNotAutoScanned() error {
	mgr := pipeline.NewAdBlockManager()
	paths, _ := mgr.LoadFunctionalSourcePaths()
	for _, p := range paths {
		if strings.Contains(p[0], "iRingo") {
			return fmt.Errorf("found iRingo")
		}
		if strings.Contains(p[0], "YouTube.Enhance") {
			return fmt.Errorf("found YouTube.Enhance")
		}
		if strings.Contains(p[0], "Apple服务增强") {
			return fmt.Errorf("found Apple服务增强")
		}
	}
	return nil
}

func testSukkaAdblockAllowlistedFull() error {
	path := filepath.Join(hub.ROOT, "rulesets/Sources", "[Sukka] Enhance Better ADBlock for Surge.sgmodule")
	if !hub.ValidateFileExists(path, "") {
		return nil
	}
	mgr := pipeline.NewAdBlockManager()
	if mgr.ResolveModuleIngestMode(path, false) != "full" {
		return fmt.Errorf("expected full mode")
	}
	return nil
}

func RunTestAdblockModuleFilter() int {
	tests := []struct {
		name string
		fn   func() error
	}{
		{"testPddModuleFullIngestKeepsAllBodyRewrite", testPddModuleFullIngestKeepsAllBodyRewrite},
		{"testWeiboAdScriptsKeptInSplit", testWeiboAdScriptsKeptInSplit},
		{"testZhihuLegacyBodyRewriteKept", testZhihuLegacyBodyRewriteKept},
		{"testRednoteScriptsFullModule", testRednoteScriptsFullModule},
		{"testUnlockWeiboScriptStillDropped", testUnlockWeiboScriptStillDropped},
		{"testPddBottomTabsBodyRewriteKept", testPddBottomTabsBodyRewriteKept},
		{"testClassifierAdblockScript", testClassifierAdblockScript},
		{"testFilenameGate", testFilenameGate},
		{"testBilibiliBundleSplitKeepsAdOnly", testBilibiliBundleSplitKeepsAdOnly},
		{"testFunctionalResolveIncludesBilibiliAdblockLocal", testFunctionalResolveIncludesBilibiliAdblockLocal},
		{"testLocalModulesOnlyFunctionalScan", testLocalModulesOnlyFunctionalScan},
		{"testAmplifyNexusNotAutoScanned", testAmplifyNexusNotAutoScanned},
		{"testSukkaAdblockAllowlistedFull", testSukkaAdblockAllowlistedFull},
	}

	failed := false
	for _, t := range tests {
		if err := runAdblockTest(t.name, t.fn); err != nil {
			fmt.Println("FAIL:", err)
			failed = true
		}
	}

	if failed {
		return 1
	}

	fmt.Println("test_adblock_module_filter: OK")
	return 0
}
